src/neighborsearch/kball_searcher.jl

```diff
@@ -0,0 +1,68 @@
+# ------------------------------------------------------------------
+# Licensed under the ISC License. See LICENSE in the project root.
+#
+# Author: @exepulveda
+# ------------------------------------------------------------------
+
+"""
+    KBallSearcher(object, k, ball; locations)
+
+A method for searching neighbors in spatial `object` inside `neighborhood`.
+Used when doing estimation combining k-nearest and radius.
+It searches first for the k-nearest by using the internal kdtree and then
+limiting the results within the `BallNeighborhood`
+"""
+struct KBallSearcher{O,K,T,M} <: AbstractBoundedNeighborSearcher
+
+  # input fields
+  object::O
+  k::Int
+  ball::BallNeighborhood{T,M}
+  locations::Vector{Int}
+
+  # state fields
+  kdtree::K
+
+end
+
+
+function KBallSearcher(object::O, k::Int, ball::BallNeighborhood{T,M}; locations=1:npoints(object)) where {O,T,M}
+  @assert 1 ≤ k ≤ length(locations) "number of neighbors must be smaller than number of locations"
+  @assert locations ⊆ 1:npoints(object) "invalid locations for spatial object"
+  kdtree = KDTree(coordinates(object, locations), ball.metric)
```

Comment:
We try to avoid accessing fields of objects directly. In this case we have `metric(ball)` defined. Could you please use that instead?

Reply:
Sure. Could we explain me the reasons?

Reply:
It is because the object `ball` has a publicly defined API that we try to use. At some point in the future if the field name changes to something else, we will still be able to call `metric(ball)`. We only access fields directly when we are implementing a method for a function that is related to the struct we are defining the behavior. For example, in this case we are implementing the `searcher` so we can access the fields directly. However, the field `ball` of the searcher has its behavior defined elsewhere in the source and so we should only rely on the public api. Makes sense?